chore(docs): New "Creating a Source Plugin" tutorial (#37538)
gatsbyjs/gatsby: docs/docs/tutorial/creating-a-source-plugin/part-4/index.mdx

---
title: "Part 4: Utilities"
---

import { LinkButton } from "gatsby-interface"
import { MdArrowForward } from "react-icons/md"

## Introduction

In the previous parts you built the foundation of your source plugin, from now on it's about adding new functionalities, improving the user experience, and making the plugin overall better.

This part of the tutorial will focus on a subset of the [Node API helpers](/docs/reference/config-files/node-api-helpers/). Feel free to revisit this document as it can also work as standalone instructions for those APIs.

By the end of this part of the tutorial, you will be able to:

- Use the `reporter` API to output structured terminal messages
- Use the `cache` API to save artifacts between runs
- Define custom errors for your plugin through `reporter.setErrorMap`
- Use the `pluginOptionsSchema` API to verify your plugin's options

## `reporter` API

> You, as a source plugin author, can interact with your users in a variety of ways. You can share information in the README, chat in issues & PRs, create guides etc. But you or rather your source plugin will interact the most with your users through the terminal when they run `gatsby develop` or `gatsby build`. Gatsby outputs what it does behind the scenes with the goal of giving the users rich information that helps them be more productive with Gatsby itself.

Reviewer: ```suggestion
You, as a source plugin author, can interact with your users in a variety of ways. You can share information in the README, chat in issues & PRs, create guides, etc. But no matter how active of a maintainer you are, it is your source plugin that will interact the most with your users through the terminal when they run `gatsby develop` or `gatsby build`. Gatsby outputs what it does behind the scenes with the goal of giving the users rich information that helps them be more productive with Gatsby itself.
```

For this, Gatsby internally uses the `reporter` API and you can use it, too! When adding logs to the terminal be mindful of keeping of them concise, informative, and to a minimum. Most Gatsby plugins don't even need to add logs (other than errors), so in doubt leave them out at the beginning and reevaluate once you have a bigger user base.

If you need information about an API, always be sure to check Gatsby's API docs, in this case the [reporter API docs](/docs/reference/config-files/node-api-helpers/#reporter). The `reporter` API is available in all Node APIs and its methods can be used like this:

```ts
// Doesn't have to only be onPluginInit, any Node API has it
export const onPluginInit = ({ reporter }) => {
  reporter.info(`Hello World!`)
}
```

> Here's a short explainer on the most important methods:

Reviewer: ```suggestion
Here's a short explanation on the most important methods:
```

    import type { GatsbyNode, SourceNodesArgs, NodeInput } from "gatsby"
    import { fetchGraphQL } from "./utils"
    import type { IAuthorInput, IPostInput, NodeBuilderInput } from "./types"
    // highlight-next-line
    import { NODE_TYPES, ERROR_CODES } from "./constants"

    export const sourceNodes: GatsbyNode[`sourceNodes`] = async (gatsbyApi) => {
      // Rest of sourceNodes

      if (errors) {
        // highlight-start
        sourcingTimer.panicOnBuild({
          id: ERROR_CODES.GraphQLSourcing,
          context: {
            sourceMessage: `Sourcing from the GraphQL API failed`,
            graphqlError: errors[0].message,
          },
        })
        // highlight-end
      }

      // Rest of sourceNodes
    }

    // Rest of file
    ```

1. As in step 3 of [Task: Output errors](#task-output-errors), add a typo to the GraphQL error again (e.g. `title2`), restart the `develop:site` script and you should see a new output for the error:

    ```shell
    ERROR #plugin_10000  PLUGIN

    Sourcing from the GraphQL API failed: Cannot query field "title2" on type "Post". Did you mean "title"?

    failed Sourcing from plugin API - 0.029s

>     ```

Reviewer: I'm pretty confident I have all the code changes applied that are covered in this section, however the error message I see in my terminal is:
<img width="784" alt="Screen Shot 2023-03-20 at 6 16 14 PM" src="https://user-images.githubusercontent.com/1790506/226477605-c99947cd-1c2a-4e23-97ad-e41c1a2f7008.png">

I just wanted to call out the discrepancy — might not be of great concern.

Author: Oh, then things are not working as expected and we should dig into this!

You need to see `ERROR #plugin_10000  PLUGIN` otherwise the custom error is not working correctly.

Reviewer: Okay I finally got around to revisiting this. I went through Part 4 again and double checked the code and I'm still not getting the custom error.

I pushed up what I'm working with locally here: https://github.com/aghreed/source-plugin-tutorial/tree/start

    A sample `gatsby-config` file:

    ```js:title=gatsby-config.js
    module.exports = {
      plugins: [
        {
          resolve: `gatsby-plugin-console-log`,
          options: { message: "Hello world" },
        },
      ],
    }
    ```

    Usage of that option in a Node API:

    ```js:title=plugins/gatsby-plugin-console-log/gatsby-node.js
    exports.onPreInit = (_, pluginOptions) => {
      console.log(
        `logging: "${pluginOptions.message}" to the console` // highlight-line
      )
    }
    ```
- A plugin can optionally define a schema (through the `pluginOptionsSchema` API) to enforce a type for each option. Gatsby will validate that the options users pass match the schema to help them correctly set up their site.

    ```js:title=plugins/gatsby-plugin-console-log/gatsby-node.js
    exports.pluginOptionsSchema = ({ Joi }) => {
      return Joi.object({
        message: Joi.string()
          .required()
          .description(`The message logged to the console.`),
      })
    }
    ```

Here are some tips when working with plugin options:

> - **Consider the options a public API contract between your users and your plugin.** Follow [SemVer](https://semver.org/) versioning when adding, modyfing, and removing options. For example, if you want to remove an option you should only do this in a major release of your plugin since it's a breaking change.

Reviewer: ```suggestion
- **Consider the options a public API contract between your users and your plugin.** Follow [SemVer](https://semver.org/) versioning when adding, modifying, and removing options. For example, if you want to remove an option you should only do this in a major release of your plugin since it's a breaking change.
```

> - **Add as little plugin options as possible.** In the beginning, use safe defaults where possible (to the best of your knowledge) and when users wish to have new options available ask yourself really hard if you should add them. Once added, you'll need to maintain the option forever (or until you remove it again).

Reviewer: ```suggestion
- **Add as few plugin options as necessary.** In the beginning, use safe defaults where possible (to the best of your knowledge) and when users wish to have new options available, seriously contemplate if you should add them. Once added, you'll need to maintain the option forever (or until you remove it again).
```

1. The boilerplate you cloned at the beginning already has an `endpoint` option defined in the site's `gatsby-config`. Go to `site/gatsby-config.ts` to inspect the code:

    ```ts:title=site/gatsby-config.ts
    import type { GatsbyConfig } from "gatsby"
    import type { IPluginOptions } from "plugin"

    // highlight-next-line
    const GRAPHQL_ENDPOINT = `http://localhost:4000/graphql`

    const config: GatsbyConfig = {
      graphqlTypegen: true,
      plugins: [
        {
          resolve: `plugin`,
          options: {
            // highlight-next-line
            endpoint: GRAPHQL_ENDPOINT,
          } satisfies IPluginOptions,
        },
        `gatsby-plugin-image`,
        `gatsby-plugin-sharp`,
        `gatsby-transformer-sharp`,
      ],
    }

    export default config
    ```

    So the plugin already receives the GraphQL endpoint through the `endpoint` option.

1. Update the TypeScript types inside `plugin/src/types.ts` so that both internally and externally the types are correct:

    ```ts:title=plugin/src/types.ts
    // Rest of types

    interface IPluginOptionsKeys {

>       // highlight-next-line

Reviewer: Since these instructions are to replace the existing interface by this name, what do you think about highlighting the whole block? Specifically lines 671 - 674.

Author: The interface should stay as-is, only the contents of it should change. I'll change the sentence for it a little bit to make that clearer.

      endpoint: string
    }

    // Rest of types
    ```

1. Open the file `plugin/src/source-nodes.ts` and validate that you can access the option. You can do this by accessing the second parameter of the `sourceNodes` function and `console.log` it. Use the `IPluginOptionsInternal` type for it:

    ```ts:title=plugin/src/source-nodes.ts
    // Rest of imports
    import type {
      IAuthorInput,
      IPostInput,
      NodeBuilderInput,
      // highlight-next-line
      IPluginOptionsInternal,
    } from "./types"

    export const sourceNodes: GatsbyNode[`sourceNodes`] = async (
      gatsbyApi,
      // highlight-next-line
      pluginOptions: IPluginOptionsInternal
    ) => {
      const { reporter, cache } = gatsbyApi

> 
>       console.log({ pluginOptions })

Reviewer: ```suggestion
      // highlight-next-line
      console.log({ pluginOptions })
```

      // Rest of sourceNodes
    }

    // Rest of file
    ```

1. Restart the `develop:site` script and you should see an output in the terminal like this:

    ```shell
    {
      pluginOptions: { plugins: [], endpoint: 'http://localhost:4000/graphql' }
    }
    ```

>     Don't worry about the `plugins` key there. Since Gatsby's plugins can also have sub-plugins, the key `plugins` is added by default to the `pluginOptions` object. The `endpoint` option successfully comes through.

Reviewer: ```suggestion
    Don't worry about the `plugins` key there. Since Gatsby's plugins can also have sub-plugins, the key `plugins` is added by default to the `pluginOptions` object. What you really should care about is that the `endpoint` option successfully comes through.
```

---
title: "Part 4: Utilities"
---

import { LinkButton } from "gatsby-interface"
import { MdArrowForward } from "react-icons/md"

## Introduction

> In the previous parts you built the foundation of your source plugin, from now on it's about adding new functionalities, improving the user experience, and making the plugin overall better.

Reviewer 2: Suggestion to change for readability:
```suggestion
In the previous parts you built the foundation of your source plugin, from now on it's about adding new functionalities, improving the user experience, and making the plugin better overall.
```

      // Types

      // highlight-next-line
      const { data, errors } = await fetchGraphQL<IApiResponse>(
        `http://localhost:4000/graphql`,
        `query FetchApi {
          posts {
            id
            slug
            title
            image {
              url
              alt
              width
              height
            }
            author
          }
          authors {
            id
            name
          }
        }
        `
      )

      // highlight-start
      if (errors) {
        // Do something...
      }
      // highlight-end

      // Rest of sourceNodes
    }

    // Rest of file
    ```

1. Inside the `errors` conditional, use `panicOnBuild` to output the error message that you're receiving from your API. `error`, `panicOnBuild`, and `panic` accept a `string`, `Error` or a structured error (you'll learn about the latter in [Task: Define custom errors](#task-define-custom-errors)).

    ```ts:title=plugin/src/source-nodes.ts
    // Imports

    export const sourceNodes: GatsbyNode[`sourceNodes`] = async (gatsbyApi) => {
      // Rest of sourceNodes

      if (errors) {
        // highlight-next-line

>         reporter.panicOnBuild(errors[0].message)

Reviewer 3: should there also be "early return" here as we probably don't want to just continue trying to create nodes if there are errors reported?

      }
      `
    )
    ```

    Restart the `develop:site` script and you should see an error in the terminal:

    ```shell
    ERROR  UNKNOWN

    Cannot query field "title2" on type "Post". Did you mean "title"?
    ```

    Yes, it works 🎉 Don't forget to remove the typo again. In [Task: Define custom errors](#task-define-custom-errors) you'll learn how to further customize this error.

    <Announcement>

    **Pro tip:** This is the error that the GraphQL API directly returns. You can see the same error if you go to `http://localhost:4000/graphql` (the GraphQL API from `api` folder) and try the query there. Get yourself familiar with how, when, and in which form your API throws errors.

    </Announcement>

You're now surfacing any errors that occur during the data fetching to your users. Use the `error`, `panicOnBuild`, and `panic` methods throughout your code to display all relevant errors to your user.

**Please note:** If necessary, add context to your error messages instead of only passing the error along. This is especially true if the error could be thrown in multiple places or if the original error is not clear enough.

### Task: Show activities and timings

> Since your source plugin is retrieving information from a third-party API the process of doing that could take a while, especially if it's fetching a lot of data. In these instances you can use the `activityTimer` method to let the user know that something is happening in the background. You shouldn't add separate `activityTimer` instances (remember: Keep the output concise and to a minimum) but instead use the `setStatus` method on `activityTimer` to update its text.

Reviewer 2: Suggestion to change for brevity:
```suggestion
Since your source plugin is retrieving information from a third-party API, that process could take a while, especially if it's fetching a lot of data. In these instances you can use the `activityTimer` method to let the user know that something is happening in the background. You shouldn't add separate `activityTimer` instances (remember: Keep the output concise and to a minimum) but instead use the `setStatus` method on `activityTimer` to update its text.
```

> You'll now exactly learn that by adding an activity for the sourcing and node creation process of your plugin.

Reviewer 2: This sentence was a little hard to parse on the first read, here's a possible suggestion for clarity: 
```suggestion
Now you'll learn exactly that, by adding an activity for the sourcing and node creation process of your plugin.
```

    </Announcement>

1. The next step is to use `cache.get` to retrieve the timestamp from the cache before the `fetchGraphQL` call. Since you won't use it for delta fetching in this tutorial, output the information in the `verbose` mode of Gatsby with `reporter.verbose`:

    ```ts:title=plugin/src/source-nodes.ts
    // Imports

    export const sourceNodes: GatsbyNode[`sourceNodes`] = async (gatsbyApi) => {
      // Rest of sourceNodes

      // highlight-next-line
      const lastFetchedDate: number = await cache.get(CACHE_KEYS.Timestamp)
      const lastFetchedDateCurrent = Date.now()

      // highlight-next-line
      reporter.verbose(`[plugin] Last fetched date: ${lastFetchedDate}`)

      // Rest of sourceNodes
    }

    // Rest of file
    ```

1. Time to test your changes! Restart the `develop:site` script but this time with the `--verbose` flag enabled:

    ```shell
    yarn develop:site --verbose
    ```

    The output got a lot busier! You should see a new log:

    ```shell
    verbose [plugin] Last fetched date: undefined
    ```

>     You're seeing this because on this first run the `lastFetcheDate` didn't exist yet and thus `undefined` is returned from the `cache.get` call. Stop the development server and restart the script. You now should see a timestamp defined that looks something like this:

Reviewer 2: Small typo:
```suggestion
    You're seeing this because on this first run the `lastFetchedDate` didn't exist yet and thus `undefined` is returned from the `cache.get` call. Stop the development server and restart the script. You now should see a timestamp defined that looks something like this:
```

    ```shell
    verbose [plugin] Last fetched date: 1677669643651
    ```

    Great, it's working!

    <Announcement>

>     **Pro tip:** If you'd save files with the `cache` API you could also inspect what is saved by going to the `.cache/caches` folder and looking for a folder named like your plugin. For this tutorial the path would be `site/.cache/caches/plugin`.

Reviewer 2: This tense is a little confusing, here's a suggestion to change to present tense:
```suggestion
    **Pro tip:** If you save files with the `cache` API you can also inspect what is saved by going to the `.cache/caches` folder and looking for a folder named like your plugin. For this tutorial the path would be `site/.cache/caches/plugin`.
```
